com.spotify/index.js

```diff
@@ -0,0 +1,662 @@
+        human_output: "No song is currently playing."}];
+      if (response === '' || response.length === 0 ) {
+        return NONE_RESULT;
+      }
+      const parsed = JSON.parse(response);
+      if (parsed.is_playing == false) {
+        return NONE_RESULT;
+      }
+      let message = "The currently playing song is '" + parsed.item.name + "'.";
+      // TODO create convenience functions to parse out groups of artists, albums, and tracks
+      // TODO use the function to expand the message to include artists.
+      return [{name: parsed.item.name, is_song_playing: true, human_output: message}];
+    });
+  }
+
+  http_put(url, data, options) {
+    return Tp.Helpers.Http.request(url, 'PUT', data, options);
+  }
+
+  http_put_default_options(url, data) {
+    return this.http_put(url, data, {
+      useOAuth2: this,
+      dataContentType: 'application/json',
+      accept: 'application/json'
+    });
+  }
+
+  http_post_default_options(url, data) {
+    console.log("post url is " + url);
+    return Tp.Helpers.Http.post(url, data, {
+      useOAuth2: this,
+      dataContentType: 'application/json',
+      accept: 'application/json'
+    });
+  }
+
+  // TODO should probably prefix each method with the Spotify API endpoint category it
+  // begins with (i.e. player, library, browse, personalization, etc.)
+  player_play_helper(data = '', options = {
+    useOAuth2: this,
+    dataContentType: 'application/json',
+    accept: 'application/json'
+  }) {
+    console.log("data is " + data)
+    return this.http_put(PLAY_URL, data, options);
+  }
+
+  do_player_pause() {
+    return this.http_put_default_options(PAUSE_URL, '');
+  }
+
+  do_player_play() {
+    console.log("Playing music...");
+    return this.player_play_helper();
+  }
+
+  do_player_next() {
+    return this.http_post_default_options(NEXT_URL, '');
+  }
+
+  do_player_previous() {
+    return this.http_post_default_options(PREVIOUS_URL, '');
+  }
+
+  do_player_shuffle({ shuffle }) {
+    // TODO figure out how to parse shuffle from natural language
+    // and set the state appropriately.
+    // change it to ENUM - then we can hardcode it into the utterance
+    shuffle = shuffle === 'on' ? 'true' : 'false';
+    console.log("shuffle: " + shuffle);
+    let shuffle_url = SHUFFLE_URL + '?state=' + shuffle;
+    return this.http_put_default_options(shuffle_url, '');
+  }
+
+  do_player_repeat({ repeat }) {
+    // TODO figure out how to parse repeat from natural language
+    // and set the state appropriately.
+    // for now, assume it's a string: track, context or off.
+    console.log("repeat: " + repeat);
+    // TODO put shuffle as a query param
+    let repeat_url = REPEAT_URL + '?state=' + repeat;
+    return this.http_put_default_options(repeat_url, '');
+  }
+
+
+
+  /* Search */
+
+  search(query, types){
+      let searchURL = SEARCH_URL + "?q=" + query.toString().replace(' ', '+') + '&type=' + types + '&market=from_token' + '&limit=1';
```

Comment:
use `encodeURIComponent(query)` rather than replacing spaces with +
encodeURIComponent will also deal with other special characters

(You can also use [querystring](https://nodejs.org/api/querystring.html) which is nicer to read)